cmd/heimdalld/main.go

```diff
@@ -223,6 +226,82 @@ func InitCmd(ctx *server.Context, cdc *codec.Codec) *cobra.Command {
 	return cmd
 }
 
+// VerifyGenesis verifies the genesis file and brings it in sync with on-chain contract
+func VerifyGenesis(ctx *server.Context, cdc *codec.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "verify-genesis",
+		Short: "Verify if the genesis matches",
+		Args:  cobra.NoArgs,
+		RunE: func(_ *cobra.Command, _ []string) error {
+			config := ctx.Config
+			config.SetRoot(viper.GetString(cli.HomeFlag))
+			helper.InitHeimdallConfig("")
+
+			// Loading genesis doc
+			genDoc, err := tmTypes.GenesisDocFromFile(filepath.Join(config.RootDir, "config/genesis.json"))
+			if err != nil {
+				return err
+			}
+
+			// get genesis state
+			var genesisState app.GenesisState
+			err = json.Unmarshal(genDoc.AppState, &genesisState)
+			if err != nil {
+				return err
+			}
+			contractCaller, err := helper.NewContractCaller()
+			if err != nil {
+				return err
+			}
+
+			// check header count
+			currentHeaderIndex, err := contractCaller.CurrentHeaderBlock()
+			if err != nil {
+				return nil
+			}
+
+			if genesisState.CheckpointData.AckCount*helper.GetConfig().ChildBlockInterval != currentHeaderIndex {
+				fmt.Println("Header Count doesn't match",
+					"ExpectedHeader", currentHeaderIndex,
+					"HeaderIndexFound", genesisState.CheckpointData.AckCount*helper.GetConfig().ChildBlockInterval)
+				return nil
+			}
+			fmt.Println("ACK count valid!","count",currentHeaderIndex)
+
+			// check all headers
+			for i, header := range genesisState.CheckpointData.Headers {
+				ackCount := uint64(i + 1)
+				root, start, end, _, _, err := contractCaller.GetHeaderInfo(ackCount * helper.GetConfig().ChildBlockInterval)
+				if err != nil {
+					return err
+				}
+				if header.StartBlock != start || header.EndBlock != end || bytes.Equal(header.RootHash.Bytes(), root.Bytes()) {
+					return fmt.Errorf("Checkpoint block doesnt match",
+						"StartExpected", header.StartBlock, "StartReceived", start,
+						"EndExpected", header.EndBlock, "EndReceived", header.EndBlock,
+						"RootHashExpected", header.RootHash.String(), "RootHashReceivd", root.String())
+				}
+				fmt.Println("Checkpoint block valid!", "start", start, "end", end, "root", root.String())
+			}
+
+			// validate validators
+			validators := genesisState.StakingData.Validators
+			for _, v := range validators {
+				val, err := contractCaller.GetValidatorInfo(v.ID)
+				if err != nil {
+					return err
+				}
+				if val.VotingPower != v.VotingPower {
```

Comment:
Please use the`GetPowerFromAmount()` method to get the amount before comparing to genesis state validator power. 

Check method: https://github.com/maticnetwork/heimdall/blob/38db2c8598ffce6594eb650565d089ff82703691/helper/util.go#L640-L647

Reply:
I think you don't it since ValidatorInfo is doing before sending it from the contract.